modules/rostests/kmtests/rtl/RtlStrSafe.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * PROJECT:         ReactOS kernel-mode tests
+ * LICENSE:         GPLv2+ - See COPYING in the top level directory
+ * PURPOSE:         Test for ntstrsafe.h functions
+ * PROGRAMMER:      Hern�n Di Pietro <hernan.di.pietro@gmail.com>
+ */
+
+#define KMT_EMULATE_KERNEL
+#include <kmt_test.h>
+#include <ntstrsafe.h>
+#include <ntdef.h>
+#include <ndk/rtlfuncs.h>
+
+#define TESTAPI static void
+
+TESTAPI
+Test_RtlUnicodeStringPrintf()
+{
+    WCHAR Buffer[1024];
+    WCHAR BufferSmall[2];
+    WCHAR BufferSmall2[7];
+    UNICODE_STRING UsString;
+    WCHAR FormatStringInts[] = L"%d %d %d";
+    WCHAR FormatStringStrs[] = L"%s %s %s";
+
+    /* No zeros (Don't assume UNICODE_STRINGS are NULL terminated) */
+
+    RtlFillMemory(Buffer, sizeof(Buffer), 0xAA);
+    RtlFillMemory(BufferSmall, sizeof(BufferSmall), 0xAA);
+    RtlFillMemory(BufferSmall2, sizeof(BufferSmall2), 0xAA);
+
+    /* STATUS_SUCCESS test */
+
+    UsString.Buffer = Buffer;
+    UsString.Length = 0;
+    UsString.MaximumLength = sizeof(Buffer);
+
+    const WCHAR Result[] = L"1 2 3";
+    ok_eq_hex(RtlUnicodeStringPrintf(&UsString, FormatStringInts, 1, 2, 3), STATUS_SUCCESS);
+    ok_eq_uint(UsString.Length, sizeof(Result) - sizeof(WCHAR));
+    ok_eq_uint(UsString.MaximumLength, sizeof(Buffer));
+    ok_eq_wchar(UsString.Buffer[0], L'1');
+    ok_eq_wchar(UsString.Buffer[1], L' ');
+    ok_eq_wchar(UsString.Buffer[2], L'2');
+    ok_eq_wchar(UsString.Buffer[3], L' ');
+    ok_eq_wchar(UsString.Buffer[4], L'3');
+    ok_eq_wchar(UsString.Buffer[5], (WCHAR) 0);
+    
+    /* STATUS_BUFFER_OVERFLOW tests */
+
+    UsString.Buffer = BufferSmall;
+    UsString.Length = 0;
+    UsString.MaximumLength = sizeof(BufferSmall);
+    
+    ok_eq_hex(RtlUnicodeStringPrintf(&UsString, FormatStringStrs, L"AAA", L"BBB", L"CCC"), STATUS_BUFFER_OVERFLOW);
```

Comment:
In general, if you write tests for possible overflows / bad behavior, think about what will happen if the function is implemented badly.
In this case, it might write outside of the buffer, and that could corrupt the stack.
So tests like these should be done on a buffer that is big enough to hold the data (and maybe something extra), where you just tell the function that it is actually smaller. (`MaximumLength`).
This way, you can validate that it does not write outside bounds (by filling the data after the `MaximumLength` before calling the function, and checking that after it returns), and in case it does write outside bounds, you will not crash.

Reply:
Sounds good.

Reply:
@hernandp Are you going to fix this?